File: peaky_finders/src/peaky_finders/serve/site_prefetch.py
"""Coordinate placement prefetch for ``peaky serve`` (PLSS + site links)."""

from __future__ import annotations

from pathlib import Path

from peaky_finders.serve.links import ServeLinksError, load_coords_site_links
from peaky_finders.serve.plss import ServePlssError, ensure_plss_for_coords
from peaky_finders.core.preset import load_preset_sites


class ServeSitePrefetchError(Exception):
    """Site placement prefetch failed."""


def load_site_placement_prefetch(
    project_dir: Path,
    lat: float,
    lon: float,
    *,
    exclude_site_slug: str | None = None,
    verbose: bool = False,
) -> dict[str, object]:
    """Return PLSS and linked peers for draft coordinates."""
    try:
        plss_result = ensure_plss_for_coords(project_dir, lat, lon, verbose=verbose)
    except ServePlssError as e:
        raise ServeSitePrefetchError(str(e)) from e

    try:
        sites = load_preset_sites(project_dir / "config.yaml")
    except ValueError as e:
        raise ServeSitePrefetchError(f"invalid preset: {e}") from e

    try:
        links = load_coords_site_links(
            project_dir,
            lat,
            lon,
            sites,
            exclude_site_slug=exclude_site_slug,
            verbose=verbose,
        )
    except ServeLinksError as e:
        raise ServeSitePrefetchError(str(e)) from e

    link_features: list[dict[str, object]] = []
    for row in links:
        slug = str(row["slug"])
        site = sites.get(slug)
        if site is None:
            continue
        link_features.append(
            {
                "type": "Feature",
                "geometry": {
                    "type": "LineString",
                    "coordinates": [
                        [float(lon), float(lat)],
                        [float(site.lon), float(site.lat)],
                    ],
                },
                "properties": {
                    "slug": slug,
                    "manual": bool(row.get("manual")),
                    "distance_km": row.get("distance_km"),
                },
            }
        )

    return {
        "lat": lat,
        "lon": lon,
        "plss": plss_result["plss"],
        "links": links,
        "links_geojson": {"type": "FeatureCollection", "features": link_features},
    }

File: crates/peaky-serve/src/site_prefetch.rs
//! Coordinate placement prefetch (draft links at a lat/lon).

use std::path::Path;
use std::sync::Arc;

use anyhow::Result;
use peaky_preset::{load_preset, Preset};
use serde_json::{json, Value};
use splatter::Session;

use crate::links::load_coords_site_links;

#[derive(Debug, thiserror::Error)]
#[error("{0}")]
pub struct SitePrefetchError(pub String);

pub async fn load_site_placement_prefetch(
    session: Arc<Session>,
    preset_path: &Path,
    lat: f64,
    lon: f64,
    exclude_site_slug: Option<&str>,
) -> Result<Value, SitePrefetchError> {
    let preset = load_preset(preset_path).map_err(|e| SitePrefetchError(format!("invalid preset: {e:#}")))?;
    let preset_for_links = preset.clone();
    let links = tokio::task::spawn_blocking({
        let session = session.clone();
        let preset_path = preset_path.to_path_buf();
        let exclude = exclude_site_slug.map(str::to_string);
        move || load_coords_site_links(&session, &preset_path, &preset_for_links, lat, lon, exclude.as_deref())
    })
    .await
    .map_err(|e| SitePrefetchError(format!("links task join: {e}")))?
    .map_err(|e| SitePrefetchError(e.0))?;

    let link_features: Vec<Value> = links
        .iter()
        .filter_map(|row| coords_link_feature(lat, lon, &preset, row))
        .collect();

    Ok(json!({
        "lat": lat,
        "lon": lon,
        "links": links,
        "links_geojson": { "type": "FeatureCollection", "features": link_features },
    }))
}

fn coords_link_feature(lat: f64, lon: f64, preset: &Preset, row: &Value) -> Option<Value> {
    let slug = row.get("slug")?.as_str()?;
    let site = preset.sites.get(slug)?;
    Some(json!({
        "type": "Feature",
        "geometry": {
            "type": "LineString",
            "coordinates": [
                [lon, lat],
                [site.loc[1], site.loc[0]],
            ],
        },
        "properties": {
            "slug": slug,
            "manual": row.get("manual").and_then(|v| v.as_bool()).unwrap_or(false),
            "distance_km": row.get("distance_km"),
        },
    }))
}
